packages/protocol/contracts/bridge/libs/LibBridgeRecall.sol

```diff
@@ -0,0 +1,160 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.20;
+
+import { AddressResolver } from "../../common/AddressResolver.sol";
+import { EtherVault } from "../EtherVault.sol";
+import { IRecallableMessageSender, IBridge } from "../IBridge.sol";
+import { IERC165Upgradeable } from
+    "@openzeppelin/contracts-upgradeable/utils/introspection/IERC165Upgradeable.sol";
+import { LibBridgeData } from "./LibBridgeData.sol";
+import { LibBridgeStatus } from "./LibBridgeStatus.sol";
+import { LibAddress } from "../../libs/LibAddress.sol";
+
+/**
+ * This library provides functions for releasing Ether (and tokens) related to
+ * message
+ * execution on the Bridge.
+ */
+
+library LibBridgeRecall {
+    using LibBridgeData for IBridge.Message;
+    using LibAddress for address;
+
+    // All of the vaults has the same interface id
+    bytes4 public constant RECALLABLE_MESSAGE_SENDER_INTERFACE_ID = 0x59dca5b0;
+
+    event MessageRecalled(
+        bytes32 indexed msgHash,
+        address to,
+        uint256 amount,
+        LibBridgeData.RecallStatus status
+    );
+
+    error B_ETHER_RELEASED_ALREADY();
+    error B_FAILED_TRANSFER();
+    error B_MSG_NOT_FAILED();
+    error B_OWNER_IS_NULL();
+    error B_WRONG_CHAIN_ID();
+
+    /**
+     * Release Ether to the message owner
+     * @dev This function releases Ether to the message owner, only if the
+     * Bridge state says:
+     * - Ether for this message has not been released before.
+     * - The message is in a failed state.
+     * @param state The current state of the Bridge
+     * @param resolver The AddressResolver instance
+     * @param message The message whose associated Ether should be released
+     * @param proof The proof data
+     */
+    function recallMessage(
+        LibBridgeData.State storage state,
+        AddressResolver resolver,
+        IBridge.Message calldata message,
+        bytes calldata proof
+    )
+        internal
+    {
+        if (message.owner == address(0)) {
```

Comment:
remove this. we shall never use `message.owner`  in recall, always use `message.sender`